src/main/java/com/pingcap/tikv/policy/RetryPolicy.java

```diff
@@ -64,7 +92,7 @@ protected boolean checkNotRecoverableException(Status status) {
                     throw new GrpcException(e);
                 }
                 try {
-                    if (recoverMethod != null && checkNotLeaderException(status)) {
+                    if (recoverMethod != null && checkNotLeaderException(e)) {
```

Comment:
and likely we need to handle error (including pd's leader switch) in unified way via handler.